File: leetcode/ds/1022. Sum of Root To Leaf Binary Numbers/1022. Sum of Root To Leaf Binary Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode*root, int prev) {
        if(root == NULL) return 0;
        if(root -> left == NULL && root -> right == NULL) return prev*2+root->val; //말단 답 반환
        return dfs(root->right, prev*2+root->val) + dfs(root->left, prev*2+root->val);
    }
    int sumRootToLeaf(TreeNode* root) {
        return dfs(root,0);
    }
};

File: leetcode/ds/1022. Sum of Root To Leaf Binary Numbers/1022. Sum of Root To Leaf Binary Numbers.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    
    pub fn sum_root_to_leaf(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        Self::dfs(root, 0)
    }

    fn dfs(root: Option<Rc<RefCell<TreeNode>>>, current: i32) -> i32 {
        if let Some(n) = root {
            let n = n.borrow();
            let new_val = current * 2 + n.val;

            if n.left.is_none() && n.right.is_none() {
                return new_val;
            }

            return Self::dfs(n.left.clone(), new_val) + Self::dfs(n.right.clone(), new_val);
        }
        0
    }
}